wgpu-hal/src/gles/adapter.rs

```diff
@@ -74,29 +157,63 @@ impl super::Adapter {
         }
     }
 
-    pub(super) unsafe fn expose(gl: glow::Context) -> crate::ExposedAdapter<super::Api> {
+    pub(super) unsafe fn expose(gl: glow::Context) -> Option<crate::ExposedAdapter<super::Api>> {
         let vendor = gl.get_parameter_string(glow::VENDOR);
         let renderer = gl.get_parameter_string(glow::RENDERER);
+        let version = gl.get_parameter_string(glow::VERSION);
+
+        let ver = Self::parse_version(&version).ok()?;
+
+        let max_texture_size = gl.get_parameter_i32(glow::MAX_TEXTURE_SIZE) as u32;
 
         let min_uniform_buffer_offset_alignment =
             gl.get_parameter_i32(glow::UNIFORM_BUFFER_OFFSET_ALIGNMENT);
-        let min_storage_buffer_offset_alignment = if super::is_webgl() {
+        let min_storage_buffer_offset_alignment = if cfg!(target_arch = "wasm32") {
             256
         } else {
             gl.get_parameter_i32(glow::SHADER_STORAGE_BUFFER_OFFSET_ALIGNMENT)
         };
 
-        crate::ExposedAdapter {
+        let limits = wgt::Limits {
+            max_texture_dimension_1d: max_texture_size,
+            max_texture_dimension_2d: max_texture_size,
+            max_texture_dimension_3d: max_texture_size,
+            max_texture_array_layers: gl.get_parameter_i32(glow::MAX_ARRAY_TEXTURE_LAYERS) as u32,
+            max_bind_groups: 4,
+            max_dynamic_uniform_buffers_per_pipeline_layout: 8,
+            max_dynamic_storage_buffers_per_pipeline_layout: 4,
+            max_sampled_textures_per_shader_stage: 16,
+            max_samplers_per_shader_stage: 16,
+            max_storage_buffers_per_shader_stage: 8,
+            max_storage_textures_per_shader_stage: 8,
+            max_uniform_buffers_per_shader_stage: 12,
+            max_uniform_buffer_binding_size: 16384,
+            max_storage_buffer_binding_size: 128 << 20,
+            max_vertex_buffers: 8,
+            max_vertex_attributes: 16,
+            max_vertex_buffer_array_stride: 2048,
+            max_push_constant_size: 0,
+        };
+
+        let features = wgt::Features::empty(); //TODO
```

Comment:
Just FYI, `glow` now exposed extensions via API (currently only available in git version). Introduced in https://github.com/grovesNL/glow/pull/165.

Reply:
This is excellent note at the right time. I definitely want to use the new API!